File: architecture-skills/object-oriented-programming-paradigm/examples/interface-segregation/interface-segregation.cs
// Interface Segregation Principle in C#.
//
// A fat IOfficeMachine interface forces SimplePrinter to stub scan()
// and fax() with NotSupportedException. The fix splits IPrinter and
// IScanner so each client depends only on the methods it calls.

using System;

// --- Violation: one fat interface every device must implement ---

public interface IOfficeMachine
{
    void PrintDocument(string path);
    void ScanDocument(string path);
    void FaxDocument(string path);
}

public sealed class SimplePrinter : IOfficeMachine
{
    public void PrintDocument(string path) => Console.WriteLine($"print {path}");

    public void ScanDocument(string path) => throw new NotSupportedException("scan");

    public void FaxDocument(string path) => throw new NotSupportedException("fax");
}

static void RunOfficeJob(IOfficeMachine machine, string path)
{
    machine.PrintDocument(path);
    foreach (var (action, fn) in new (string, Action)[]
             {
                 ("scan", () => machine.ScanDocument(path)),
                 ("fax", () => machine.FaxDocument(path)),
             })
    {
        try { fn(); }
        catch (NotSupportedException) { Console.WriteLine($"{action} attempt: NotSupported"); }
    }
}

// --- Fix: small client-specific interfaces ---

public interface IPrinter
{
    void PrintDocument(string path);
}

public interface IScanner
{
    void ScanDocument(string path);
}

public sealed class BasicPrinter : IPrinter
{
    public void PrintDocument(string path) => Console.WriteLine($"print {path}");
}

public sealed class FlatbedScanner : IScanner
{
    public void ScanDocument(string path) => Console.WriteLine($"scan {path}");
}

public sealed class MultiFunctionDevice : IPrinter, IScanner
{
    public void PrintDocument(string path) => Console.WriteLine($"print {path}");
    public void ScanDocument(string path) => Console.WriteLine($"scan {path}");
}

static void PrintJob(IPrinter printer, string path) => printer.PrintDocument(path);
static void ScanJob(IScanner scanner, string path) => scanner.ScanDocument(path);

public static class Program
{
    public static void Main()
    {
        Console.WriteLine("-- Violation: SimplePrinter forced to stub scan/fax --");
        RunOfficeJob(new SimplePrinter(), "document.pdf");

        Console.WriteLine("\n-- Fix: clients depend only on what they need --");
        PrintJob(new BasicPrinter(), "document.pdf");
        ScanJob(new FlatbedScanner(), "page.tif");
        PrintJob(new MultiFunctionDevice(), "report.pdf");
    }
}

File: architecture-skills/object-oriented-programming-paradigm/examples/interface-segregation/interface-segregation.php
<?php
// Interface Segregation Principle in PHP.
//
// A fat IOfficeMachine interface forces SimplePrinter to stub scanDocument
// and faxDocument with exceptions. The fix splits IPrinter and IScanner
// so each client depends only on the methods it calls.

declare(strict_types=1);

// --- Violation: one fat interface every device must implement ---

interface IOfficeMachine
{
    public function printDocument(string $path): void;

    public function scanDocument(string $path): void;

    public function faxDocument(string $path): void;
}

final class SimplePrinter implements IOfficeMachine
{
    public function printDocument(string $path): void
    {
        echo "print {$path}\n";
    }

    public function scanDocument(string $path): void
    {
        throw new RuntimeException('scan');
    }

    public function faxDocument(string $path): void
    {
        throw new RuntimeException('fax');
    }
}

function runOfficeJob(IOfficeMachine $machine, string $path): void
{
    $machine->printDocument($path);
    foreach (['scan' => 'scanDocument', 'fax' => 'faxDocument'] as $action => $method) {
        try {
            $machine->{$method}($path);
        } catch (RuntimeException) {
            echo "{$action} attempt: NotSupported\n";
        }
    }
}

// --- Fix: small client-specific interfaces ---

interface IPrinter
{
    public function printDocument(string $path): void;
}

interface IScanner
{
    public function scanDocument(string $path): void;
}

final class BasicPrinter implements IPrinter
{
    public function printDocument(string $path): void
    {
        echo "print {$path}\n";
    }
}

final class FlatbedScanner implements IScanner
{
    public function scanDocument(string $path): void
    {
        echo "scan {$path}\n";
    }
}

final class MultiFunctionDevice implements IPrinter, IScanner
{
    public function printDocument(string $path): void
    {
        echo "print {$path}\n";
    }

    public function scanDocument(string $path): void
    {
        echo "scan {$path}\n";
    }
}

function printJob(IPrinter $printer, string $path): void
{
    $printer->printDocument($path);
}

function scanJob(IScanner $scanner, string $path): void
{
    $scanner->scanDocument($path);
}

function main(): void
{
    echo "-- Violation: SimplePrinter forced to stub scan/fax --\n";
    runOfficeJob(new SimplePrinter(), 'document.pdf');

    echo "\n-- Fix: clients depend only on what they need --\n";
    printJob(new BasicPrinter(), 'document.pdf');
    scanJob(new FlatbedScanner(), 'page.tif');
    printJob(new MultiFunctionDevice(), 'report.pdf');
}

main();
